lib/statespace_sse.h

```diff
@@ -0,0 +1,214 @@
+// limitations under the License.
+
+#ifndef STATESPACE_SSE_H_
+#define STATESPACE_SSE_H_
+
+#include <xmmintrin.h>
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <cstdint>
+#include <functional>
+
+#include "statespace.h"
+#include "util.h"
+
+namespace qsim {
+
+// Routines for state-vector manipulations.
+// State is a vectorized sequence of four real components followed by four
+// imaginary components. Four single-precison floating numbers can be loaded
+// into an SSE register.
+template <typename ParallelFor>
+struct StateSpaceSSE : public StateSpace<ParallelFor, float> {
+  using Base = StateSpace<ParallelFor, float>;
+  using State = typename Base::State;
+  using fp_type = typename Base::fp_type;
+
+  StateSpaceSSE(unsigned num_qubits, unsigned num_threads)
+      : Base(num_qubits, num_threads,
+             2 * std::max(uint64_t{4}, uint64_t{1} << num_qubits)),
+        num_qubits_(num_qubits) {}
+
+  void SetAllZeros(State& state) const {
+    __m128 val0 = _mm_setzero_ps();
+
+    auto f = [](unsigned n, unsigned m, uint64_t i,
+                const __m128& val0, State& state) {
+      _mm_store_ps(state.get() + 8 * i + 0, val0);
+      _mm_store_ps(state.get() + 8 * i + 4, val0);
+    };
+
+    ParallelFor::Run(Base::num_threads_, Base::raw_size_ / 8, f, val0, state);
+  }
+
+  // Uniform superposition.
+  void SetStateUniform(State& state) const {
+    uint64_t size = uint64_t{1} << num_qubits_;
+
+    __m128 val0 = _mm_setzero_ps();
+    __m128 valu = _mm_set1_ps(fp_type{1} / std::sqrt(size));
+
+    auto f = [](unsigned n, unsigned m, uint64_t i,
+                const __m128& val0, const __m128& valu, State& state) {
+      _mm_store_ps(state.get() + 8 * i + 0, valu);
+      _mm_store_ps(state.get() + 8 * i + 4, val0);
+    };
+
+    ParallelFor::Run(Base::num_threads_, Base::raw_size_ / 8, f,
+                     val0, valu, state);
+  }
+
+  // |0> state.
+  void SetStateZero(State& state) const {
+    SetAllZeros(state);
+    state.get()[0] = 1;
+  }
+
+  static std::complex<fp_type> GetAmpl(const State& state, uint64_t i) {
+    uint64_t p = (8 * (i / 4)) + (i % 4);
+    return std::complex<fp_type>(state.get()[p], state.get()[p + 4]);
+  }
+
+  static void SetAmpl(
+      const State& state, uint64_t i, const std::complex<fp_type>& ampl) {
+    uint64_t p = (8 * (i / 4)) + (i % 4);
+    state.get()[p + 0] = std::real(ampl);
+    state.get()[p + 4] = std::imag(ampl);
+  }
+
+  static void SetAmpl(const State& state, uint64_t i, fp_type re, fp_type im) {
+    uint64_t p = (8 * (i / 4)) + (i % 4);
+    state.get()[p + 0] = re;
+    state.get()[p + 4] = im;
+  }
+
+  double Norm(const State& state) const {
+    using Op = std::plus<double>;
+
+    auto f = [](unsigned n, unsigned m, uint64_t i,
+                const State& state) -> double {
+      __m128 re = _mm_load_ps(state.get() + 8 * i);
+      __m128 im = _mm_load_ps(state.get() + 8 * i + 4);
```

Comment:
SSE doesn't have the high precision fmadd instructions that FMA and AVX2 use. If people want good precision at higher qubit numbers would it be worth using pd here instead of ps ?

Reply:
Here and in other places, I sum four single precision numbers and convert the result to double precision. The problem would arise if I didn't convert to double precision. I think it's not worth using pd here as the simulator intrinsically uses single precision and we'll not gain any extra precision here.